feat: hook reactions in read only channel and fix issues

jrainville opened the pull request:
Fixes #13523

Status-go PR: https://github.com/status-im/status-go/pull/4900

This hooks the feature that let's users react in a view only channel, but only if the setting is enabled by admins.

It also fixes a couple small issues, 2 of them in status-go.

Known issues:
- It's very small, but the context menu doesn't look very good. That's because I had to temper with the QMl myself and there is no design or component right now for just emojis. Let me know if I'm wrong. Otherwise, it can be fixed in another issue/PR

[reactions.webm](https://github.com/status-im/status-desktop/assets/11926403/6a7ea6cc-f063-4a1e-a171-3193bac9c5fd)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | fdbdc48c | [#1](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-13913/1/) | 2024-03-08 22:15:27 | ~5 min | `tests/nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-13913/1//consoleText) |

jrainville commented on src/app_service/service/chat/dto/chat.nim:
```diff
@@ -416,3 +423,12 @@ proc communityChannelUuid*(self: ChatDto): string =
     return ""
   return self.id[self.communityId.len .. ^1]
+
+proc updateMissingFields*(chatToUpdate: var ChatDto, oldChat: ChatDto) =
```
I moved this here because it is needed by the chat service too, so now both community and chat services can use it.

jrainville commented on src/app_service/service/chat/service.nim:
```diff
@@ -165,13 +165,16 @@ QtObject:
               self.events.emit(SIGNAL_CHAT_MEMBERS_CHANGED, ChatMembersChangedArgs(chatId: chatDto.id, members: chatDto.members))
-            self.updateOrAddChat(chatDto)
+            if isCommunityChat:
+              updatedChat.updateMissingFields(oldChat)
```
This was needed, because otherwise we lost the community fields like `viewersCanPostReactions` (they reseted to `false`)

jrainville commented on ui/imports/shared/stores/PermissionsStore.qml:
```diff
@@ -17,10 +17,6 @@ QtObject {
 
-    function setViewOnlyCanAddReaction(chatId, checked) {
-        //TODO: backend implementation
-    }
```
I removed those because that field isn't updated independently, it's updated using the EditChat API

jrainville commented on ui/imports/shared/views/chat/ChatContextMenuView.qml:
```diff
@@ -41,8 +42,6 @@ StatusMenu {
     signal updateGroupChatDetails(string chatId, string groupName, string groupColor, string groupImage)
 
-    signal createCommunityChannel(string chatId, string newName, string newDescription, string newEmoji, string newColor)
-    signal editCommunityChannel(string chatId, string newName, string newDescription, string newEmoji, string newColor, string newCategory)
```
Those were not used :scissors:

jrainville commented on ui/imports/shared/views/chat/MessageView.qml:
I had to do a couple changes in this file and QML so that the context menu with just Emojis appeared. It's not perfect visually, but I didn't want to waste too much time on it. I think we should open another issue after that to polish the UI

jrainville reviewed (commented):
I added some comments to explain some of the code changes

jrainville commented on src/app_service/service/community/service.nim:
```diff
@@ -641,10 +634,20 @@ QtObject:
-            updatedChat.updateMissingFields(chat)
+          if chat.name != prevChat.name or chat.description != prevChat.description or chat.color != prevChat.color or
+              chat.emoji != prevChat.emoji or chat.viewersCanPostReactions != prevChat.viewersCanPostReactions:
+            var updatedChat = chat
```
Note here that I'm no longer using `updatedChats`, but the `chat` directly here, because `updatedChats` was missing a LOT of the fields, and not just community fields. It was also missing its ID and communityID.

I don't think it should cause a lot of problems to use `chat` directly. I tested editing the channel in multiple ways and they all worked.

alexjba reviewed (commented):
<img width="1112" alt="Screenshot 2024-03-11 at 21 33 08" src="https://github.com/status-im/status-desktop/assets/47811206/8210efc1-bedb-4e92-b591-9ebac56490e2">

The emojis are no longer available on hover for group chats.
Looks good to me otherwise.

jrainville commented:
> The emojis are no longer available on hover for group chats. Looks good to me otherwise.

Nice find. I fixed that in status-go.

I also fixed a crash I introduced.

jrainville commented:
The status-go PR is now merged. I rebased and updated the vendor. 

@alexjba @igor-sirotin you can review :D

igor-sirotin commented on src/app/modules/main/chat_section/item.nim:
```diff
@@ -35,6 +35,9 @@ type
     loaderActive: bool
     locked: bool
     requiresPermissions: bool
+    canPostMessages: bool
+    canPostReactions: bool
```
I wonder if this partially duplicates this:
https://github.com/status-im/status-desktop/blob/3ef21897fd70a49bb9bbc5fbc94b1012383e45ae/src/app/modules/main/chat_section/item.nim#L39-L40

It's in https://github.com/status-im/status-desktop/pull/13876

jrainville replied:
Hmm, good question. I guess `canPostMessages` is the equivalent of  `viewAndPostPermissionsSatisfied`.

The `canPostReactions` is necessary for my PR and it's definitely different.

I added the `canPostMessages` because it was given by status-go, but we don't use it for now, so maybe it's worth removing to avoid confusion? cc @endulab

igor-sirotin commented on src/app_service/service/community/service.nim:
```diff
@@ -641,10 +634,20 @@ QtObject:
+            # Currently, status-go only sends canPostReactions on app start (getChannelGroups)
+            # so here, we need to imply it. If viewersCanPostReactions is true, then everyone can post reactions
+            # admins can also always post reactions
+            if chat.viewersCanPostReactions == true or
+                (chat.viewersCanPostReactions == false and community.memberRole != MemberRole.None):
+              updatedChat.canPostReactions = true
+            elif chat.viewersCanPostReactions == false and community.memberRole == MemberRole.None:
+              updatedChat.canPostReactions = false
```
```suggestion
            if chat.viewersCanPostReactions or
                (!chat.viewersCanPostReactions and community.memberRole != MemberRole.None):
              updatedChat.canPostReactions = true
            elif !chat.viewersCanPostReactions and community.memberRole == MemberRole.None:
              updatedChat.canPostReactions = false
```

jrainville replied:
Yeah, I used those explicit conditions when developing because trying to get the right conditions was a pain so it was easier to understand for me, but then I forgot to simplify haha.

BTW, it's `not` instead of `!` in Nim :wink:

igor-sirotin commented on ui/imports/shared/views/chat/MessageView.qml:
```diff
@@ -910,7 +910,7 @@ Loader {
                 quickActions: [
                     Loader {
-                        active: d.addReactionAllowed && delegate.hovered && !delegate.hideQuickActions
+                        active: d.addReactionAllowed && delegate.hovered
```
hmm, why don't we want to keep `&& !delegate.hideQuickActions`? 🤔

jrainville replied:
I didn't want to tamper too much with the QML code because it's not my jurisdiction, but here it's because we want to still show the Emoji reacitons when doing a right-click on a message in a read-only channel.

The code is not really made for that, so that's why it's a bit weird in places.

igor-sirotin replied:
Looks like it's important to keep here.
Guessing from  the definition:
```qml
readonly property bool hideQuickActions: root.isChatBlocked ||
                  root.placeholderMessage ||
                  root.isInPinnedPopup ||
                  root.editModeOn ||
                  !root.rootStore.mainModuleInst.activeSection.joined
```

we should hide the _and disable the loaders_ in a few cases. 
`isInPinnedPopup` is also covered by `addReactionAllowed`. But all other cases doesn't seem to be covered.

I understand that `isChatBlocked` doesn't consider `canPostReactions` at all. 
So we should probably review the set of properties here so that it fit our needs.

igor-sirotin commented on ui/imports/shared/views/chat/MessageContextMenuView.qml:
```diff
@@ -23,6 +23,7 @@ StatusMenu {
     property string myPublicKey: ""
     property bool amIChatAdmin: false
     property bool disabledForChat: false
+    property bool showOnlyEmojiReactions: false
```
"show only" means that as soon as we set this to `true`, no other options will appear.
But that wouldn't be true when `disabledForChat` is false.

```suggestion
    property bool forceEnableEmojiReactions: false
```

igor-sirotin commented:
Thank for this, looks cool!

alexjba reviewed (approved):
Nice work!
